### PR TITLE
Vscatters for v65 Hexagon target

### PR DESCRIPTION
**Pull request by aankit-ca**
Scatter + Scatter_acc for v65 Hexagon target.
@dsharletg @pranavb-ca @dpalermo PTAL

### CODE REVIEW

**Comment by aankit-ca** [comment]
At present, the PR is missing a correctness test for scatter.

File: src/CodeGen_Hexagon.cpp [hunk 1]
```diff
@@ -1933,6 +1933,38 @@ void CodeGen_Hexagon::visit(const Call *op) {
             value = builder->CreateCall(fn, args);
         }
         return;
+    } else if (op->is_intrinsic() && (op->name == "scatter" || op->name == "scatter_acc")) {
+        internal_assert(op->args.size() == 4);
+        internal_assert(op->type.bits() != 8);
```
[the remark this answers is not in the document]
Comment (aankit-ca) [reply src/CodeGen_Hexagon.cpp]: Is this not good?
internal_assert(op->type.bits() != 8);

File: src/HexagonOptimize.cpp [hunk 2]
```diff
@@ -2104,8 +2122,120 @@ class ScatterGatherGenerator : public IRMutator2 {
...
+        return scatter;
+    }
+};
+
+// Scatter-Gather instructions on Hexagon are asynchronous and hence require a
+// scatter-release store followed by a vector load from the same address. This
+// stalls the pipeline untill all previous scatter-gather operations have
+// finished. The operations are not ordered with respect to load and store
+// operations as well.
+class SyncronizationBarriers : public IRMutator2 {
```
[the remark this answers is not in the document]
Comment (aankit-ca) [reply src/HexagonOptimize.cpp]: The code identifies the following scenarios:

- Scatter-gather
- Scatter-scatter
- Gather-Scatter
- Gather-gather
- Scatter-Load
- Scatter-Store
- Gather-Store

The producer-consumer relation won't exist for any of the above cases except for scatter-load.
The code deals with update definitions as well.

File: src/HexagonOptimize.cpp [hunk 3]
```diff
@@ -2104,8 +2122,120 @@ class ScatterGatherGenerator : public IRMutator2 {
...
+        Expr buffer = Variable::make(Handle(), op->name);
+        Expr index = mutate(cast(ty.with_code(Type::Int), ty.bytes() * op->index));
+        value = mutate(value);
+        Stmt scatter = Evaluate::make(Call::make(ty, intrinsic,
+                              {buffer, size-1, index, value}, Call::Intrinsic));
+        return scatter;
+    }
+};
+
+// Scatter-Gather instructions on Hexagon are asynchronous and hence require a
```
[the remark this answers is not in the document]
Comment (aankit-ca) [reply src/HexagonOptimize.cpp]: Yes Dillon. A gather-store, gather-gather, gather-scatter all require scatter-release->load to order the operation.

**Comment by aankit-ca** [comment]
@dsharletg PTAL

File: src/CodeGen_Hexagon.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (dsharletg) [reply src/CodeGen_Hexagon.cpp]: What about 64 bits? Or something else wacky? It's minor, but this code only works with 16 and 32 bits, nothing else. Pranav's rewrite reflects that.

**Comment by dsharletg** [comment]
Going to clone this for testing, but reminding you that the build bots don't currently test v65, so it will only check against breakages of existing code.

**Comment by aankit-ca** [comment]
@dsharletg @pranavb-ca Should we merge this?

**Comment by dsharletg** [comment]
The test branch looks slightly stale, but it did pass. I don't think any of the recent changes would break anything, and travis did check this branch. Going to merge it.

**Pull request merged**